codeintel: Automatically fetch and index LSIF-referenced NPM dependencies

varungandhi-src opened the pull request:
Builds on top of https://github.com/sourcegraph/sourcegraph/pull/28057 and https://github.com/sourcegraph/sourcegraph/pull/22984/ .

TODOs:
- [x] Add automated tests.
- [x] Figure out where to put the scheme constants. Earlier, I ran into an import cycle error when I did it carelessly. This has some explicit TODOs in the code.

Manual testing steps:
  1. Start with `sg start enterprise-codeintel`.
  2. Set `npmPackages` to enabled under experimental features under site-admin.
  3. Refresh and add NPM Dependencies as a code host.
  4. Add `"got@12.0.0"` under `"dependencies"` in the code host settings. 
  5. [Turn on auto-indexing](https://docs.sourcegraph.com/code_intelligence/how-to/enable_auto_indexing) and set up a policy. Make sure that repos with URLs matching `npm/*` will be covered.
  6. ```
      npm pack got@12.0.0
      tar -xzvf got-12.0.0.tar.gz
      cd package
      git init && git add . && git commit -m "Temp."
      echo '{"compilerOptions":{"allowJs":true}}' > tsconfig.json
      npm install --ignore-scripts
      lsif-tsc -p .
      # Assuming you've got SRC_ACCESS_TOKEN and SRC_ENDPOINT set up in the environment
      src lsif upload -repo=npm/got -root=. -commit=bb23d1f0cb5d634d1da32804860810f8f2d5c176
      ````
  7. `got` and its dependencies should be available for search shortly (e.g. try searching for `interface`).

efritz commented:
Reviewed all code up to this point - no blocking comments 🎉 .  Tap me if you'd like another review after some tests are added (some around the database are blockers, plus any additional tests where you think they would be most useful).

varungandhi-src commented:
@efritz thanks for the quick review. I've added some tests in the latest commit. 3 commits before that are just cleanups.

I wasn't entirely sure what kinds of tests to add, so I've added some mimicking the existing ones with explicit dependencies. LMK if you have other ideas for what kinds of tests would be valuable to add here.

efritz commented on enterprise/cmd/worker/internal/codeintel/indexing/dependency_sync_scheduler_test.go:
```diff
@@ -8,6 +8,7 @@ import (
 
 	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore"
 	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/shared"
+	dbstore2 "github.com/sourcegraph/sourcegraph/internal/codeintel/stores/dbstore"
```
I'd call this `ossdbstore` or the other one `enterprisedbstore`

efritz commented on internal/codeintel/stores/dbstore/repos.go:
```diff
@@ -96,6 +103,71 @@ func scanJVMDependencyRepo(rows *sql.Rows, queryErr error) (dependencies []JVMDe
 	return dependencies, nil
 }
 
+func (s *Store) GetNPMDependencyRepos(ctx context.Context, filter GetNPMDependencyReposOpts) (repos []NPMDependencyRepo, err error) {
```
I see that neither `GetNPMDependencyRepos` nor `GetJVMDependencyRepos` methods have tests. This should be a huge part that gets exercised in all code paths as possible (as we have no static guarantee that we're forming appropriate SQL).

For a sample test to copy, see [TestGetUploads](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@0cdf3b914fb99f2a7fbf124ab0822d22e46e118e/-/blob/enterprise/internal/codeintel/stores/dbstore/uploads_test.go?L275:6). Would be great to get a test for both, but at least one for the new code should make it in pre-merge.

varungandhi-src replied:
I've added a test case for this function in a recent commit. I've marked a few places with `[Question]` for things I wasn't sure about.

1. Are database IDs always 1 or higher?
2. What exactly are the considerations around using code from `enterprise/*` in a non-enterprise directly? Do those apply to tests as well? I've copied a bit of code for now.
3. I wrote some code for checking that an insert succeeded, I was wondering if there was a more concise/idiomatic way to do the same.

efritz replied:
1. Yes, but we shouldn't necessarily assume that the IDs will _start_ at one or be gapless (transaction rollback doesn't rollback sequence numbers, etc).
2. We have a check in CI to disallow any imports of enterprise code from OSS code. Any import of the `enterprise` package (Go or Typescript) must only be imported from something also under that directory. See other inline comments about copying vs sharing in the case of tests.
3. I think this is spot on 🎉 , see inline comments for more context.

efritz commented on internal/repos/npm_packages.go:
```diff
@@ -48,7 +57,42 @@ func (s *NPMPackagesSource) ListRepos(ctx context.Context, results chan SourceRe
+		return
+	}
+	totalDBFetched, totalDBResolved, lastID := 0, 0, 0
+	for {
```
This sequence of logic could also benefit from a test that mocks the dbstore. Seems like this is an easy place for miscounts and off-by-one errors to easily sneak in (I see the JVM version has no tests either).

varungandhi-src replied:
Added a test case for `ListRepo`.

efritz commented on internal/codeintel/stores/dbstore/repos_test.go:
```diff
@@ -23,3 +27,94 @@ func TestRepoName(t *testing.T) {
+	for _, depStr := range dependencies {
+		// [Varun][Question] Is there a better way to check that running the query didn't cause any errors?
+		dep, err := reposource.ParseNPMDependency(depStr)
+		assert.Nil(t, err)
```
If you want the error to be fatal, you can use `require.Nil(t, err)`, which requires a symmetric import of `"github.com/stretchr/testify/require"`.

efritz commented on internal/codeintel/stores/dbstore/repos_test.go:
```diff
@@ -23,3 +27,94 @@ func TestRepoName(t *testing.T) {
+	}
+}
+
+// [Varun][Question] Should we directly call 'InsertCloneableDependencyRepo' from
```
It is *completely* fine to duplicate a bit of code in order to set up the database in the way a test expects. This is part of the ways I designed the basestore API as I did, as it allows you perform raw queries in test without having to do a lot of jumping through hoops to get a handle to the database. See [this example](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@0fdc44a374723c6904d6887607e28f8a2af3eced/-/blob/enterprise/internal/codeintel/stores/dbstore/migration/committed_at_test.go?L75) of inline queries in a test.

efritz commented on internal/codeintel/stores/dbstore/repos_test.go:
```diff
@@ -23,3 +27,94 @@ func TestRepoName(t *testing.T) {
+const insertCloneableDependencyRepoQuery = `
+INSERT INTO lsif_dependency_repos (scheme, name, version)
+VALUES (%s, %s, %s)
+ON CONFLICT DO NOTHING
```
This can be simplified for this test - we don't need to check for conflicts or return anything for a count, so the query can be as simple as `INSERT INTO (...) VALUES (...)` (and could even be inlined in the test).

efritz commented on internal/codeintel/stores/dbstore/repos_test.go:
```diff
@@ -23,3 +27,94 @@ func TestRepoName(t *testing.T) {
+		}
+		sort.Strings(depStrs)
+		sort.Strings(testCase.matches)
+		assert.Equal(t, testCase.matches, depStrs)
```
BTW, thank you for using this package. I'd like us to use this package more heavily in our tests instead of using direct comparisons via operators and `t.Fail(...)`.

Having more code using it is good motivation to begin the transition, so keep it up!

Strum355 replied:
And here I have been doing direct comparisons for ages (or google-cmp) because thats the way you've been doing everywhere :robpike:

Strum355 commented on cmd/gitserver/server/vcs_syncer_npm_packages.go:
```diff
@@ -176,14 +178,40 @@ func (s *NPMPackagesSyncer) packageDependencies(ctx context.Context, repoUrlPath
+			log15.Error("failed to parse npm package", "package", dbDep.Package, "message", err)
+			continue
+		}
+		if *repoPackage == *parsedDbPackage {
```
Any particular reason these two are pointers? In the JVM equivalent we have them as value types

varungandhi-src replied:
The parsing functions return pointers which are `nil` when parsing fails. I guess they could return a zero value like the JVM functions, I just found it a bit odd to do that because the zero value is invalid. In case you somehow forgot to check the error, you'd get an immediate crash with `nil` instead of a failure somewhere later with an invalid zero value.

efritz replied:
Zero values are meant to be meaningful in Go (in general).

varungandhi-src replied:
Yes, I'm painfully aware of the idea, but there is no meaning I can ascribe here to a package with an empty name. I actually ran into this exact problem when debugging a little earlier which was quite frustrating; the validation and serialization/deserialization functions in `reposource/npm_packages.go` are defenses to avoid that exact problem in the future.

```go
// An NPM package of the form (@scope/)?name.
//
// The fields are kept private to reduce risk of not handling the empty scope
// case correctly.
type NPMPackage struct {
```

Strum355 reviewed (approved):
Nothing blocking on my end :+1:

varungandhi-src commented:
I've factored out the changes into separate commits. Most are no-ops, there are a couple which add tests. Feel free to skip the in-between ones.

I've added a FIXME for deduplicating the list of repos returned; I will fix that in a follow-up PR.
